File: ubh_jaspr/lib/components/email_capture.dart
import 'package:jaspr/jaspr.dart';
import 'package:jaspr/dom.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';

@client
class EmailCapture extends StatefulComponent {
  const EmailCapture({super.key});

  @override
  State<EmailCapture> createState() => _EmailCaptureState();
}

class _EmailCaptureState extends State<EmailCapture> {
  bool isPopupOpen = true;
  bool isWidgetOpen = false;
  String email = '';
  String statusMessage = '';
  bool isSuccess = false;
  bool isLoading = false;

  Future<void> _submitEmail() async {
    final trimmedEmail = email.trim();
    if (trimmedEmail.isEmpty) {
      setState(() {
        isSuccess = false;
        statusMessage = 'PLEASE ENTER A VALID EMAIL.';
      });
      return;
    }

    setState(() {
      isLoading = true;
      statusMessage = 'SUBMITTING...';
    });

    try {
      final response = await http.post(
        Uri.parse('https://subscribe-loz23viwea-uc.a.run.app'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'email': trimmedEmail}),
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        setState(() {
          isSuccess = true;
          statusMessage = 'WELCOME TO THE ROSTER.';
          email = '';
        });

        // Auto-close popup after a short delay
        if (isPopupOpen) {
          Future.delayed(const Duration(seconds: 2), () {
            if (mounted) {
              setState(() {
                isPopupOpen = false;
                statusMessage = '';
              });
            }
          });
        }
      } else {
        setState(() {
          isSuccess = false;
          statusMessage = 'SUBMISSION FAILED. TRY AGAIN.';
        });
      }
    } catch (e) {
      setState(() {
        isSuccess = false;
        statusMessage = 'ERROR CONNECTING TO SERVER.';
      });
    } finally {
      setState(() {
        isLoading = false;
      });
    }
  }

  @override
  Component build(BuildContext context) {
    if (isPopupOpen) {
      return div(
        classes: 'fixed inset-0 z-50 bg-black/80 flex items-center justify-center p-4',
        [
          div(
            classes: 'relative bg-[#0d0d0d] border border-[#d4af37]/30 rounded-xl max-w-md w-full p-8 text-center shadow-2xl',
            [
              // X Close Button
              button(
                classes: 'absolute top-4 right-4 text-zinc-400 hover:text-white transition-colors duration-200 cursor-pointer text-xl font-bold bg-transparent border-0',
                onClick: () {
                  setState(() {
                    isPopupOpen = false;
                    statusMessage = '';
                  });
                },
                [text('×')]
              ),
              
              h2(
                classes: 'text-3xl md:text-4xl font-extrabold font-["Space_Grotesk"] text-[#d4af37] tracking-wider mb-2 uppercase',
                [text('JOIN THE HUNT')]
              ),
              
              p(
                classes: 'text-zinc-400 text-sm mb-6 max-w-xs mx-auto font-["Manrope"]',
                [text('Subscribe for exclusive UBH updates, early drops, and priority studio booking.')]
              ),
              
              input(
                type: InputType.email,
                classes: 'w-full bg-[#141414] border border-[#262626] focus:border-[#d4af37] text-white px-4 py-3 rounded-lg outline-none mb-4 transition-colors font-["Manrope"] placeholder-zinc-600',
                attributes: {'placeholder': 'ENTER YOUR EMAIL', 'required': 'true'},
                value: email,
                onInput: (val) {
                  email = val.toString();
                }
              ),
              
              button(
                classes: 'w-full bg-[#d4af37] hover:bg-[#b8952b] text-black font-extrabold py-3 px-6 rounded-lg uppercase tracking-wider transition-colors font-["Space_Grotesk"] cursor-pointer disabled:opacity-55',
                attributes: isLoading ? {'disabled': 'true'} : {},
                onClick: _submitEmail,
                [text(isLoading ? 'SUBMITTING...' : 'JOIN THE HUNT')]
              ),
              
              if (statusMessage.isNotEmpty)
                p(
                  classes: 'text-sm mt-4 font-["Manrope"] ${isSuccess ? "text-green-500" : "text-red-500"}',
                  [text(statusMessage)]
                )
            ]
          )
        ]
      );
    } else {
      // Floating Widget bottom-6 right-6
      return div(
        classes: 'fixed bottom-6 right-6 z-40 flex flex-col items-end gap-3',
        [
          if (isWidgetOpen)
            div(
              classes: 'bg-[#0d0d0d] border border-[#d4af37]/30 rounded-lg p-5 w-80 shadow-2xl text-left flex flex-col gap-3 font-["Manrope"]',
              [
                h3(
                  classes: 'text-sm font-bold text-[#d4af37] uppercase tracking-wider font-["Space_Grotesk"]',
                  [text('GET UBH EXCLUSIVES')]
                ),
                p(
                  classes: 'text-xs text-zinc-400',
                  [text('Enter your email for drop alerts:')]
                ),
                input(
                  type: InputType.email,
                  classes: 'w-full bg-[#141414] border border-[#262626] focus:border-[#d4af37] text-white text-sm px-3 py-2 rounded outline-none placeholder-zinc-600',
                  attributes: {'placeholder': 'YOUR EMAIL', 'required': 'true'},
                  value: email,
                  onInput: (val) {
                    email = val.toString();
                  }
                ),
                button(
                  classes: 'w-full bg-[#d4af37] hover:bg-[#b8952b] text-black text-xs font-bold py-2 rounded uppercase tracking-wider transition-colors cursor-pointer disabled:opacity-55',
                  attributes: isLoading ? {'disabled': 'true'} : {},
                  onClick: _submitEmail,
                  [text(isLoading ? 'SUBMITTING...' : 'SUBSCRIBE')]
                ),
                if (statusMessage.isNotEmpty)
                  p(
                    classes: 'text-xs font-medium ${isSuccess ? "text-green-500" : "text-red-500"}',
                    [text(statusMessage)]
                  )
              ]
            ),
          button(
            classes: 'w-14 h-14 bg-[#d4af37] hover:bg-[#b8952b] text-black rounded-full flex items-center justify-center shadow-lg transition-transform hover:scale-105 duration-200 cursor-pointer text-2xl border-0',
            onClick: () {
              setState(() {
                isWidgetOpen = !isWidgetOpen;
                statusMessage = ''; // Clear status message on toggle
              });
            },
            [text(isWidgetOpen ? '×' : '✉️')]
          )
        ]
      );
    }
  }
}

File: components/email_capture.py
from fasthtml.common import *

def EmailCaptureModal():
    return Div(
        Div(
            Div(
                # Close button
                Button(
                    "✕",
                    onclick="document.getElementById('email-capture-modal').style.display='none'",
                    cls="absolute top-4 right-4 text-neutral-400 hover:text-white text-lg font-bold bg-transparent border-0 cursor-pointer"
                ),
                # Insignia
                Div(
                    Span("⚡", cls="text-2xl mb-1 block"),
                    Span("JOIN THE HUNT", cls="text-xs font-heading font-bold tracking-[0.4em] text-[#D4AF37] uppercase block"),
                    H3("VIP ROSTER ACCESS", cls="text-2xl md:text-3xl font-black font-heading tracking-tight text-white uppercase mt-1"),
                    P("Subscribe for instant alerts on Rap Funxtion tickets, studio lock drops, and limited-edition collective uniform drops.", cls="text-neutral-400 text-xs md:text-sm mt-2 max-w-xs mx-auto leading-relaxed"),
                    cls="text-center mb-6"
                ),
                # Form
                Form(
                    Div(
                        Input(
                            type="email",
                            name="email",
                            placeholder="ENTER YOUR EMAIL ADDRESS",
                            required=True,
                            id="modal-email-input",
                            cls="input-dark w-full text-center text-xs md:text-sm font-heading tracking-wider mb-3"
                        ),
                        Button(
                            "CLAIM ACCESS",
                            type="submit",
                            cls="btn-gold w-full text-xs font-heading font-black tracking-widest py-3"
                        ),
                        Div(id="modal-subscribe-status", cls="text-center text-xs mt-3"),
                        cls="flex flex-col"
                    ),
                    onsubmit="""
                        event.preventDefault();
                        const emailInput = document.getElementById('modal-email-input');
                        const statusDiv = document.getElementById('modal-subscribe-status');
                        statusDiv.innerHTML = '<span class=\"text-[#D4AF37]\">ENCRYPTING & SUBMITTING...</span>';
                        fetch('/subscribe', {
                            method: 'POST',
                            headers: {'Content-Type': 'application/x-www-form-urlencoded'},
                            body: 'email=' + encodeURIComponent(emailInput.value)
                        }).then(res => res.json()).then(data => {
                            if (data.success) {
                                statusDiv.innerHTML = '<span class=\"text-emerald-400 font-bold font-heading\">WELCOME TO THE ROSTER.</span>';
                                emailInput.value = '';
                                setTimeout(() => { document.getElementById('email-capture-modal').style.display='none'; }, 2000);
                            } else {
                                statusDiv.innerHTML = '<span class=\"text-rose-400\">' + (data.error || 'Submission failed.') + '</span>';
                            }
                        }).catch(e => {
                            statusDiv.innerHTML = '<span class=\"text-rose-400\">Network error.</span>';
                        });
                    """
                ),
                cls="relative bg-[#0D0D0D] border border-[#D4AF37]/40 rounded-2xl max-w-md w-full p-8 shadow-2xl gold-glow"
            ),
            cls="min-h-screen flex items-center justify-center p-4"
        ),
        id="email-capture-modal",
        cls="fixed inset-0 z-50 bg-black/85 backdrop-blur-sm hidden"
    )

def EmailCaptureWidget():
    return Div(
        # Floating toggle button
        Div(
            Div(
                # Pop-up drawer
                Div(
                    H4("GET UBH EXCLUSIVES", cls="text-xs font-heading font-bold text-[#D4AF37] tracking-wider uppercase mb-1"),
                    P("Enter your email for priority studio and show drop alerts:", cls="text-[11px] text-neutral-400 mb-3"),
                    Form(
                        Input(
                            type="email",
                            name="email",
                            placeholder="YOUR EMAIL",
                            required=True,
                            id="widget-email-input",
                            cls="input-dark w-full text-xs py-2 px-3 mb-2 font-body"
                        ),
                        Button(
                            "SUBSCRIBE",
                            type="submit",
                            cls="btn-gold w-full text-xs py-2 tracking-widest font-heading font-black"
                        ),
                        Div(id="widget-subscribe-status", cls="text-xs mt-2 text-center"),
                        onsubmit="""
                            event.preventDefault();
                            const emailInput = document.getElementById('widget-email-input');
                            const statusDiv = document.getElementById('widget-subscribe-status');
                            statusDiv.innerHTML = '<span class=\"text-[#D4AF37]\">Submitting...</span>';
                            fetch('/subscribe', {
                                method: 'POST',
                                headers: {'Content-Type': 'application/x-www-form-urlencoded'},
                                body: 'email=' + encodeURIComponent(emailInput.value)
                            }).then(res => res.json()).then(data => {
                                if (data.success) {
                                    statusDiv.innerHTML = '<span class=\"text-emerald-400 font-bold\">Subscribed!</span>';
                                    emailInput.value = '';
                                } else {
                                    statusDiv.innerHTML = '<span class=\"text-rose-400\">' + (data.error || 'Failed') + '</span>';
                                }
                            }).catch(e => {
                                statusDiv.innerHTML = '<span class=\"text-rose-400\">Error</span>';
                            });
                        """
                    ),
                    id="email-widget-drawer",
                    cls="hidden mb-3 bg-[#0D0D0D] border border-[#D4AF37]/40 rounded-xl p-5 w-72 md:w-80 shadow-2xl text-left"
                ),
                Button(
                    Span("✉️", cls="text-xl"),
                    onclick="""
                        const drawer = document.getElementById('email-widget-drawer');
                        drawer.classList.toggle('hidden');
                    """,
                    cls="w-12 h-12 md:w-14 md:h-14 bg-[#D4AF37] hover:bg-[#FFD700] text-black rounded-full flex items-center justify-center shadow-lg transition-transform hover:scale-105 duration-200 cursor-pointer border-0 ml-auto"
                ),
                cls="flex flex-col items-end"
            ),
            cls="fixed bottom-6 right-6 z-40"
        ),
        id="email-widget-container"
    )
